src/tests/universal-nrp-e2e/README.md

```diff
@@ -0,0 +1,281 @@
+                                   - Clean cloud-init flags to reset cloud-init to initial-state
+       -l Log Directory:         Directory used to place output logs
+       -d Debug Mode:            VM stays up for debugging (Default: false)
+```
+### VM Image Sources
+Its preferable for images to be "Cloud" images (contain cloud-init) as it makes tooling work out-of-box and does not require creating your own image of the distro and getting all the necessary packages/dependencies installed.
+
+Here are a few locations where cloud images can be obtained:
+ - [Ubuntu Cloud Images](https://cloud-images.ubuntu.com/)
+ - [Debian Cloud Images](https://cloud.debian.org/images/cloud/)
+ - [CentOS Cloud Images](https://cloud.centos.org/centos/)
+ - [Rocky Linux Images](https://rockylinux.org/download) - also contains Cloud images
+ - [Oracle Cloud Images](https://yum.oracle.com/oracle-linux-templates.html)
+
+## Generalizing a VM Image
+Generalizing an image is useful if its going to be reused for re-running tests on it. The tooling provides the ability to provision the VM, install dependencies, cleans logs, bash history and clears the cloud-init boot-flags so it can be reprovisioned appropriately. Once an image is generalized you should copy it somewhere so it can be copied and reused. The debug mode (`-d`) is useful here as it allows you to ssh into the VM once it has been provisioned so you may install any additional dependencies and/or fix any issues with the image.
+
+You typically want to generalize images that do not just work out-of-box in order to keep a "Golden Master" image that just works to avoid having to reprovision VM images when you want to re-run tests on a particular VM image.
+
+## Example - Using CentOS-7 an EOL distribution with errors
+This example will generalize an older distro image, where we will hit many issues. This example will demonstrate some common problems can can occur using older distributions where they are no longer being maintained and you must build and install some of the needed dependencies.
+
+This example demonstrates:
+ - EOL distribution whose package mirrors are out of date and must be updated
+ - Install dependencies that exist in the package-mirrors
+ - Install dependencies that require building/installing sources
+
+Using the _CentOS-7_ distribution below.
+```sh
+# Download the image
+wget https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-2211.qcow2
+# Add some additional space just to ensure satisfactory space for installing tools/dependencies
+qemu-img resize CentOS-7-x86_64-GenericCloud-2211.qcow2 +2G
+# Start VM Image with extra memory (4096mb) generalization and debug flag
+./StartVMTest.sh -i CentOS-7-x86_64-GenericCloud-2211.qcow2 -m 4096 -g -d
+```
+The VM will boot and show the console, so any errors can be resolved afterwards, the utility will provide ssh login information in order for you to go and resolve any issues.
```

Comment:
Do we need to be concerned with how the SSH information is displayed?

Reply:
These are ephemeral keys, single-use for every VM instance.